---
title: "Fungal Analysis"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r Load Elements}
library(phyloseq)
library(tidyverse)
library(ggpubr)
library(vegan)
library(Maaslin2)

# Load data
df = read.csv("Data Files/SEED_abundances.csv")

# Taxonomy
t = df %>% select(contains('Subsystem'),Function) %>% 
  mutate(OTU = paste(Subsystem.Level.1,Subsystem.Level.2,Subsystem.Level.3, Function,sep='|')) %>% 
  mutate_all(function(x)str_replace_all(x,"'",'.')) %>% 
  column_to_rownames('OTU') %>% as.matrix %>% tax_table()

# Abundance
o = df %>% select(-all_of(colnames(t))) %>% as.matrix %>% `rownames<-`(rownames(t)) %>%
  otu_table(taxa_are_rows = T)

# Sample data
s = read.csv('Data Files/metadata.csv') %>% `rownames<-`(.$Sample) %>% 
  mutate(Mouse = str_remove(Sample,'_D0') %>% str_remove('_D28'))

# Phyloseq object
ps = phyloseq(sample_data(s),o,t)
ps@sam_data$Group = factor(ps@sam_data$Group, levels = c('CON Day 0','LM Day 0','CON Day 28','LM Day 28'))

# Rarefied
set.seed(421)
ps.rar = ps %>% transform_sample_counts(function(x)round(x,0)) %>% rarefy_even_depth()
```

## Alpha Diversity

```{r}
otu = ps.rar@otu_table %>% as.matrix %>% as.data.frame
# Use Vegan package to calculate diversity
df_alpha <- diversity(t(otu), index='shannon', MARGIN = 1, base = exp(1)) %>%
  as.data.frame()
df_alpha$Sample = rownames(df_alpha)
df_alpha = df_alpha %>% left_join(s %>% select(Diet, Day) %>% rownames_to_column('Sample'))
names(df_alpha)[1] = 'shannon'
  
# Plot results - samples are paired
ggplot(df_alpha, aes(x=Diet, y=shannon)) + 
      geom_boxplot(outlier.shape = NA) +
      geom_jitter(aes(color=Day),  height = 0, width = 0.2, size=3) + 
      theme_classic(base_size = 16) +
      stat_compare_means(vjust=0.5, label.y = 8, paired=T, size = 5)+
      ylab('Shannon Diversity') +  xlab('') +
      facet_wrap(~Day) +
      theme(legend.position = 'none')
ggsave('Plots/Functional/P1_alphadiversity_SEED.png',height = 3.5, width = 6)
```

## Beta Diversity

```{r Calculate Beta diversity}
# Calculate the distance matrix:
beta_dist <- vegdist(t(otu), index = 'bray')
# Scaling results with NMDS
mds <- metaMDS(beta_dist)
# Extract PCoA axes and format
mds_data <- as.data.frame(mds$points)
mds_data$`Sample` <- rownames(mds_data)
mds_data <- mds_data %>% left_join(s %>% select(Diet, Day) %>% rownames_to_column('Sample')) %>% 
  mutate(beta_type = 'Bray-Curtis')

# Plot results
mds_data$Group = paste(mds_data$Diet, mds_data$Day)
mds_data$Group = factor(mds_data$Group, 
                        levels = c('CON Day 0','LM Day 0','CON Day 28','LM Day 28'))
ggplot(mds_data, aes(x = MDS1, y = MDS2, color = Group)) +
    geom_point(size = 2) + 
    stat_ellipse(size = 1) +
    theme_classic(base_size = 20) + 
    xlab('PCoA 1') + ylab('PCoA 2') +
    ggtitle('Bray-Curtis Beta Diversity')
  
ggsave('Plots/Functional/P1_betadiversity_SEED.png',height = 5, width = 7)
```

```{r Calculate Beta statistics}
# CENTROID DISTANCES (PERMANOVA)~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Format datasets
count_mat = otu %>% t() %>% as.data.frame() %>% rownames_to_column('Sample') %>% 
  left_join(mds_data %>% select(Sample, Diet, Day, Group)) %>% 
  mutate(Group = as.character(Group))
 
# CON D0 vs D28
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'CON')]
cm_con = otu %>% t %>% .[str_detect(rownames(.),'CON'),]
beta_dist_con <- vegdist(cm_con, index = 'bray')
pathotype.adonis.con <- ((adonis(beta_dist_con ~ temp_group))$aov.tab)[1,6] # p value
      
# LM D0 vs D28
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'LM')]
cm_lm = otu %>% t %>% .[str_detect(rownames(.),'LM'),]
beta_dist_lm <- vegdist(cm_lm, index = 'bray')
pathotype.adonis.lm <- ((adonis(beta_dist_lm ~ temp_group))$aov.tab)[1,6] # p value
      
# D0 CON vs LM
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'D0')]
cm_d0 = otu %>% t %>% .[str_detect(rownames(.),'D0'),]
beta_dist_d0 <- vegdist(cm_d0, index = 'bray')
pathotype.adonis.d0 <- ((adonis(beta_dist_d0 ~ temp_group))$aov.tab)[1,6] # p value

# D28 CON vs LM
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'D28')]
cm_d28 = otu %>% t %>% .[str_detect(rownames(.),'D28'),]
beta_dist_d28 <- vegdist(cm_d28, index = 'bray')
pathotype.adonis.d28 <- ((adonis(beta_dist_d28 ~ temp_group))$aov.tab)[1,6] # p value

# CONCATENATE STATISTICS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

beta_stats = tibble(Stat = rep('Centroid Distances',4),
                    Comparison = c('CON D0 vs D28','LM D0 vs D28',
                                   'D0 CON vs LM','D28 CON vs LM'),
                    pval = c(pathotype.adonis.con, pathotype.adonis.lm,
                             pathotype.adonis.d0, pathotype.adonis.d28)) %>% 
  mutate(`p adj` = p.adjust(pval)) %>% 
  mutate(`Beta Measure` = 'Bray-Curtis') %>% 
  select(`Beta Measure`, Stat, everything()) %>% 
  mutate(pval = round(pval,4),`p adj` = round(`p adj`,4))

write.csv(beta_stats,'Results Files/Functional/S1_beta_diversity_statistics_SEED.csv',row.names = F)

rm(beta_dist,beta_dist_con,beta_dist_d0,beta_dist_d28,beta_dist_lm,
   cm_con,cm_lm,cm_d0,cm_d28,
   pathotype.adonis.con,pathotype.adonis.d0,pathotype.adonis.d28,
   pathotype.adonis.lm,
   count_mat, mds, mds_data, temp_group)
```


```{r Ordered Barplot Function}
plot_ordered_bar<-function (physeq, x = "Diet", y = "Abundance",
                            fill,wrap='Day') {
  require(ggplot2)
  require(phyloseq)
  require(plyr)
  require(grid)
  
  # physeq = phylum3;x='Day';y='Abundance';fill='Phylum';wrap='Diet'
  
  bb <- psmelt(physeq)
  
  samp_names <- aggregate(bb$Abundance, by=list(bb$Sample), FUN=sum)[,1]
  .e <- environment()
  bb_order <- rev(c('Misc/Other',unique(bb[,fill])[unique(bb[,fill]) != 'Misc/Other'] %>% 
    sort() %>% rev()))
  bb[,fill]<- factor(bb[,fill], bb_order) #fill to total=1
  
  
  bb<- bb[order(bb[,fill]),] # genus to fill
  p = ggplot(bb, aes_string(x = x, y = y, fill = fill), 
             environment = .e, ordered = FALSE) +
    geom_bar(stat = "identity", 
             position = "stack", 
             color = "black") +
    theme_classic(base_size = 18) +
    guides(fill = guide_legend(override.aes = list(colour = NULL), reverse=F)) + 
    theme(legend.key = element_rect(colour = "black")) + 
    theme(legend.position = "right",
          legend.text = element_text(size = 16)) + ylab('Rel. Abundance')

  if (!is.null(wrap)) { p <- p + facet_wrap(wrap) }
  return(p)
}
```

```{r Make Grouped Barplots}
# SL1~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# New phyloseq object:
phylum <- ps %>% microbiome::transform('compositional') %>% 
  tax_glom(taxrank="Subsystem.Level.1", NArm=FALSE)
# Must be at least 1%
phylum2 = prune_taxa(taxa_sums(phylum)/nsamples(phylum) > 0.01, phylum) %>% 
  subset_taxa(Subsystem.Level.1 != 'Miscellaneous') %>% 
  merge_samples('Group')
# Proportions for each group now add up to 10 because n=10
temp = 10-phyloseq::sample_sums(phylum2) # Everything under 1% & unassigned reads
temp2 = otu_table(phylum2) %>% as.data.frame() %>% 
  cbind('Misc/Other' = temp) %>% as.matrix()
temp2 = temp2/10 # Convert to average proportions /1
temp3 = tax_table(phylum2) %>% as.data.frame() %>% # Adjust the taxonomy table
  rbind('Misc/Other' = c(rep('Misc/Other',4))) %>% as.matrix()
sample_temp = sample_data(phylum2) %>% as.data.frame()
sample_temp$Diet[str_detect(row.names(sample_temp), 'CON')] = 'CON'
sample_temp$Diet[str_detect(row.names(sample_temp), 'LM')] = 'LM'
sample_temp$Day[str_detect(row.names(sample_temp), '0')] = 'Day 0'
sample_temp$Day[str_detect(row.names(sample_temp), '28')] = 'Day 28'
phylum3 <- phyloseq(otu_table(temp2, taxa_are_rows = F), sample_data(sample_temp), tax_table(temp3))

plot_ordered_bar(phylum3,fill = 'Subsystem.Level.1') +xlab('')

ggsave('Plots/Functional/P1_barplots_SL1.png',height = 8, width = 10)

# SL2~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# New phyloseq object:
phylum <- ps %>% microbiome::transform('compositional') %>% 
  tax_glom(taxrank="Subsystem.Level.2", NArm=FALSE)
# Must be at least 1%
phylum2 = prune_taxa(taxa_sums(phylum)/nsamples(phylum) > 0.01, phylum) %>% 
  subset_taxa(Subsystem.Level.2 != '-') %>% 
  merge_samples('Group')
# Proportions for each group now add up to 10 because n=10
temp = 10-phyloseq::sample_sums(phylum2) # Everything under 1% & unassigned reads
temp2 = otu_table(phylum2) %>% as.data.frame() %>% 
  cbind('Misc/Other' = temp) %>% as.matrix()
temp2 = temp2/10 # Convert to average proportions /1
temp3 = tax_table(phylum2) %>% as.data.frame() %>% # Adjust the taxonomy table
  rbind('Misc/Other' = c(rep('Misc/Other',4))) %>% as.matrix()
sample_temp = sample_data(phylum2) %>% as.data.frame()
sample_temp$Diet[str_detect(row.names(sample_temp), 'CON')] = 'CON'
sample_temp$Diet[str_detect(row.names(sample_temp), 'LM')] = 'LM'
sample_temp$Day[str_detect(row.names(sample_temp), '0')] = 'Day 0'
sample_temp$Day[str_detect(row.names(sample_temp), '28')] = 'Day 28'
phylum3 <- phyloseq(otu_table(temp2, taxa_are_rows = F), sample_data(sample_temp), tax_table(temp3))

plot_ordered_bar(phylum3,fill = 'Subsystem.Level.2') +xlab('')

ggsave('Plots/Functional/P1_barplots_SL2.png',height = 8, width = 15)
```

# Run Maaslin2
```{r}
temp = ps %>% subset_samples(Day=='Day 28')
# Eukaryotic species
maas.sp = Maaslin2(
    input_data = temp@otu_table %>% as.matrix, 
    input_metadata = s, 
    output = 'temp', 
    min_abundance = 1e-5,
    min_prevalence = 0.3,
    max_significance = 0.05,
    normalization = 'TSS',
    transform = 'AST',
    fixed_effects = 'Diet',
    standardize = F,
    plot_heatmap = F,
    plot_scatter = F,
    reference = 'Diet,CON')
gc()

temp = ps %>% tax_glom('Subsystem.Level.3') %>% subset_samples(Day=='Day 28')
# Eukaryotic genus
maas.g = Maaslin2(
    input_data = temp@otu_table %>% as.matrix, 
    input_metadata = s, 
    output = 'temp', 
    min_abundance = 1e-5,
    min_prevalence = 0.3,
    max_significance = 0.05,
    normalization = 'TSS',
    transform = 'AST',
    fixed_effects = 'Diet',
    standardize = F,
    plot_heatmap = F,
    plot_scatter = F,
    reference = 'Diet,CON')
gc()

temp = ps %>% tax_glom('Subsystem.Level.2') %>% subset_samples(Day=='Day 28')
# Eukaryotic family
maas.f = Maaslin2(
    input_data = temp@otu_table %>% as.matrix, 
    input_metadata = s, 
    output = 'temp', 
    min_abundance = 1e-5,
    min_prevalence = 0.3,
    max_significance = 0.05,
    normalization = 'TSS',
    transform = 'AST',
    fixed_effects = 'Diet',
    standardize = F,
    plot_heatmap = F,
    plot_scatter = F,
    reference = 'Diet,CON')
gc()

temp = ps %>% tax_glom('Subsystem.Level.1') %>% subset_samples(Day=='Day 28')
# Eukaryotic species
maas.p = Maaslin2(
    input_data = temp@otu_table %>% as.matrix, 
    input_metadata = s, 
    output = 'temp', 
    min_abundance = 1e-5,
    min_prevalence = 0.3,
    max_significance = 0.05,
    normalization = 'TSS',
    transform = 'AST',
    fixed_effects = 'Diet',
    standardize = F,
    plot_heatmap = F,
    plot_scatter = F,
    reference = 'Diet,CON')
gc()
```

```{r}
res = maas.sp$results %>% mutate(Level = "Function") %>%
  rbind(maas.g$results %>% mutate(Level = "SL3")) %>%
  rbind(maas.f$results %>% mutate(Level = "SL2")) %>%
  rbind(maas.p$results %>% mutate(Level = "SL1")) %>% 
  mutate(feature = str_replace_all(feature,'[^[:alnum:]|^_]','.'))

t2 = ps@tax_table %>% as.matrix %>% as.data.frame %>% rownames_to_column('OTU') %>% mutate(OTU = str_replace_all(OTU,'[^[:alnum:]|^_]','.') %>% str_replace_all('[|]','.'))

res = res %>%
  dplyr::rename(OTU = feature) %>% 
  left_join(t2) %>% 
  select(OTU,Level,everything())

res$Function[res$Level != 'Function'] = NA
res$Subsystem.Level.3[!(res$Level %in% c('SL3','Function'))] = NA
res$Subsystem.Level.2[!(res$Level %in% c('SL2','SL3','Function'))] = NA

sig = res %>% filter(qval<0.01)

writexl::write_xlsx(list('All Results' = res, 'All Diet Sig' = sig),
                    'Results Files/Functional/Maaslin2_SEED_D28.xlsx')
saveRDS(list(Function=maas.sp,SL3=maas.g,SL2=maas.f,SL1=maas.p),
        'Results Files/Functional/Maaslin2_SEED_D28.rds')
```

```{r LFC}
lfc.f = ps %>% subset_samples(Day=='Day 28') %>% microbiome::transform('compositional') %>% 
  merge_samples('Group') %>% psmelt() %>% 
  select(OTU:Abundance,contains('Subsystem'),Function) %>% 
  pivot_wider(names_from = Sample, values_from = Abundance) %>% 
  mutate(lfc = log2(`LM Day 28`/`CON Day 28`)) %>% 
  mutate(lfc = ifelse(is.infinite(lfc) & lfc<0,
                      0-(1.2*max(abs(.$lfc[!is.infinite(.$lfc)]),na.rm=T)),
                      ifelse(is.infinite(lfc) & lfc>0,
                             1.2*max(abs(.$lfc[!is.infinite(.$lfc)]),na.rm=T),lfc)))

lfc.3 = ps %>% subset_samples(Day=='Day 28') %>% tax_glom("Subsystem.Level.3") %>% 
  microbiome::transform('compositional') %>% 
  merge_samples('Group') %>% psmelt() %>% 
  select(OTU:Abundance,contains('Subsystem')) %>% 
  pivot_wider(names_from = Sample, values_from = Abundance) %>% 
  mutate(lfc = log2(`LM Day 28`/`CON Day 28`)) %>% 
  mutate(lfc = ifelse(is.infinite(lfc) & lfc<0,
                      0-(1.2*max(abs(.$lfc[!is.infinite(.$lfc)]),na.rm=T)),
                      ifelse(is.infinite(lfc) & lfc>0,
                             1.2*max(abs(.$lfc[!is.infinite(.$lfc)]),na.rm=T),lfc)))

lfc.2 = ps %>% subset_samples(Day=='Day 28') %>% tax_glom("Subsystem.Level.2") %>%
  microbiome::transform('compositional') %>% 
  merge_samples('Group') %>% psmelt() %>% 
  select(OTU:Abundance,contains('Subsystem')) %>% 
  pivot_wider(names_from = Sample, values_from = Abundance) %>% 
  mutate(lfc = log2(`LM Day 28`/`CON Day 28`)) %>% 
  mutate(lfc = ifelse(is.infinite(lfc) & lfc<0,
                      0-(1.2*max(abs(.$lfc[!is.infinite(.$lfc)]),na.rm=T)),
                      ifelse(is.infinite(lfc) & lfc>0,
                             1.2*max(abs(.$lfc[!is.infinite(.$lfc)]),na.rm=T),lfc)))

lfc.1 = ps %>% subset_samples(Day=='Day 28') %>% tax_glom("Subsystem.Level.1") %>%
  microbiome::transform('compositional') %>% 
  merge_samples('Group') %>% psmelt() %>% 
  select(OTU:Abundance,contains('Subsystem')) %>% 
  pivot_wider(names_from = Sample, values_from = Abundance) %>% 
  mutate(lfc = log2(`LM Day 28`/`CON Day 28`)) %>% 
  mutate(lfc = ifelse(is.infinite(lfc) & lfc<0,
                      0-(1.2*max(abs(.$lfc[!is.infinite(.$lfc)]),na.rm=T)),
                      ifelse(is.infinite(lfc) & lfc>0,
                             1.2*max(abs(.$lfc[!is.infinite(.$lfc)]),na.rm=T),lfc)))
```

```{r Volcano Plots}
# Functional
temp = res %>% filter(Level=='Function') %>% 
  left_join(lfc.f %>% select(contains('Subsystem'),Function,lfc)) %>% 
  mutate(diff = ifelse((lfc>1 & qval<0.05),'UP',ifelse((lfc<(-1) & qval<0.05),'DOWN','NO'))) %>% 
  mutate(diff = factor(.$diff, levels = c('NO','UP','DOWN')))

gf = ggplot(temp, aes(x=lfc, y=-log10(qval), col=diff)) + geom_point() + 
  theme_classic(base_size=16) +
  geom_vline(xintercept=c(-1,1), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red") +
  scale_color_manual(values=c("black", "blue", "red")) +
  xlab('Log 2 Fold Change (LM/CON)') +
  theme(legend.position = 'none') +
  ggtitle('Individual Functions')

# Functional
temp = res %>% filter(Level=='SL3') %>% 
  left_join(lfc.3 %>% select(contains('Subsystem'),lfc)) %>% 
  mutate(diff = ifelse((lfc>1 & qval<0.05),'UP',ifelse((lfc<(-1) & qval<0.05),'DOWN','NO'))) %>% 
  mutate(diff = factor(.$diff, levels = c('NO','UP','DOWN')))

g3 = ggplot(temp, aes(x=lfc, y=-log10(qval), col=diff)) + geom_point() + 
  theme_classic(base_size=16) +
  geom_vline(xintercept=c(-1,1), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red") +
  scale_color_manual(values=c("black", "blue", "red")) +
  xlab('Log 2 Fold Change (LM/CON)') +
  theme(legend.position = 'none') +
  ggtitle('Subsystem Level 3')

# Functional
temp = res %>% filter(Level=='SL2') %>% 
  left_join(lfc.2 %>% select(contains('Subsystem'),lfc)) %>% 
  mutate(diff = ifelse((lfc>1 & qval<0.05),'UP',ifelse((lfc<(-1) & qval<0.05),'DOWN','NO'))) %>% 
  mutate(diff = factor(.$diff, levels = c('NO','UP','DOWN')))

g2 = ggplot(temp, aes(x=lfc, y=-log10(qval), col=diff)) + geom_point() + 
  theme_classic(base_size=16) +
  geom_vline(xintercept=c(-1,1), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red") +
  scale_color_manual(values=c("black", "blue", "red")) +
  xlab('Log 2 Fold Change (LM/CON)') +
  theme(legend.position = 'none') +
  ggtitle('Subsystem Level 2')

# Functional
temp = res %>% filter(Level=='SL1') %>% 
  left_join(lfc.1 %>% select(contains('Subsystem'),lfc)) %>% 
  mutate(diff = ifelse((lfc>1 & qval<0.05),'UP',ifelse((lfc<(-1) & qval<0.05),'DOWN','NO'))) %>% 
  mutate(diff = factor(.$diff, levels = c('NO','UP','DOWN')))

g1 = ggplot(temp, aes(x=lfc, y=-log10(qval), col=diff)) + geom_point() + 
  theme_classic(base_size=16) +
  geom_vline(xintercept=c(-1,1), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red") +
  scale_color_manual(values=c("black", "blue", "red")) +
  xlab('Log 2 Fold Change (LM/CON)') +
  theme(legend.position = 'none') +
  ggtitle('Subsystem Level 1')

ggarrange(plotlist = lapply(list(gf,g3,g2,g1),function(x) x+ylim(0,13)+xlim(-15,15)), ncol = 2, nrow = 2,legend='none')

ggsave('Results Files/Functional/Maaslin2_SEED_Volcano.jpeg',height = 6, width = 7.5)
```