---
title: "6. Figure 4 with CARD Database"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r Load Data}
library(phyloseq)
library(tidyverse)
library(ggpubr)
```

```{r Load SEED Data}
# Load data
df = read.csv("Data Files/SEED_abundances.csv")

# Taxonomy
t = df %>% select(contains('Subsystem'),Function) %>% 
  mutate(OTU = paste(Subsystem.Level.1,Subsystem.Level.2,Subsystem.Level.3, Function,sep='|')) %>% 
  mutate_all(function(x)str_replace_all(x,"'",'.')) %>% 
  column_to_rownames('OTU') %>% as.matrix %>% tax_table()

# Abundance
o = df %>% select(-all_of(colnames(t))) %>% as.matrix %>% `rownames<-`(rownames(t))

# Convert to TPM
cs = colSums(o)/1e6 # Denominator
o2 = (t(o)/cs) %>% t()

# Sample data
s = read.csv('Data Files/metadata.csv') %>% `rownames<-`(.$Sample) %>% 
  mutate(Mouse = str_remove(Sample,'_D0') %>% str_remove('_D28'))

# Phyloseq object
ps = phyloseq(sample_data(s),otu_table(o2,taxa_are_rows=T),t)
ps@sam_data$Group = factor(ps@sam_data$Group, levels = c('CON Day 0','LM Day 0','CON Day 28','LM Day 28'))

cols = c('CON' = '#FC746E','LM' = '#01BEC3')

# SEED Maaslin2 results
maas = readxl::read_xlsx('Results Files/Functional/Maaslin2_SEED_D28.xlsx',sheet = 'All Results')
```

```{r Prep Oxidative stress}
to_test = c('Oxidative stress','Oxygen stress response / Human gut microbiome',
            'Protection from Reactive Oxygen Species','Regulation of Oxidative Stress Response')
temp = ps %>% subset_taxa(Subsystem.Level.3 %in% to_test) %>% 
  tax_glom('Subsystem.Level.3') %>% 
  psmelt()

pval = maas %>% filter(Level=='SL3', Subsystem.Level.3 %in% to_test) %>% 
  select(Subsystem.Level.3,qval) %>% 
  mutate(symb = ifelse(qval>0.05,'',ifelse(qval>0.01,"*",
                      ifelse(qval>0.001,'**',ifelse(qval>0.0001,'***','****'))))) %>% 
  filter(symb !='')

temp = temp %>% filter(Subsystem.Level.3 %in% pval$Subsystem.Level.3)

lm_order = temp %>% filter(Day == 'Day 28', Diet=='LM') %>% 
  group_by(Sample,Subsystem.Level.3) %>% 
  summarize(Abundance = sum(Abundance)) %>% ungroup() %>% 
  group_by(Subsystem.Level.3) %>% 
  summarize(Abundance = mean(Abundance)) %>% ungroup() %>% 
  arrange(-Abundance)

# Filter for day 28
temp = temp %>% filter(Day == 'Day 28')
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Correlate pathobionts with ARG mechanisms

```{r Prep Pathobionts}
# Pathobionts of interest
# Genus level:
path.genus = c('Streptococcus','Staphylococcus','Enterococcus','Escherichia','Shigella','Haemophilus','Campylobacter','Klebsiella')
# Species level:
# NOTE: phocaeicola vulgatus = bacteroides vulgatus
path.species = c('Phocaeicola vulgatus', 'Prevotella copri','Bacteroides fragilis')
  
# Load using a function so that ps isn't overwritten
load_path = function(){
  source('0. Initialize Bacterial Data.R')
  ps.rel = ps %>% microbiome::transform('compositional')
  
  # Maaslin stats
  maas = readxl::read_xlsx('Results Files/Bacteriome/Maaslin2_longitudinal_bacteria.xlsx',sheet='All Results')

  # Extract otu data into table
  p = ps.rel %>% tax_glom('Phylum') %>% subset_taxa(Phylum == 'Proteobacteria') %>% 
    psmelt() %>% select(-OTU) %>% select(-Kingdom)
  f = ps.rel %>% tax_glom('Family') %>% subset_taxa(Family == 'Enterobacteriaceae') %>% 
    psmelt() %>% select(-OTU) %>% select(-c(Kingdom:Order))
  g = ps.rel %>% subset_taxa(Genus %in% path.genus) %>% tax_glom('Genus') %>% 
    psmelt() %>% rename(Pathobiont = Genus) %>% select(-OTU) %>% select(-c(Kingdom:Family))
  sp = ps.rel %>% subset_taxa(Species %in% path.species) %>% tax_glom('Species') %>% 
    psmelt() %>% rename(Pathobiont = Species) %>% select(-OTU) %>% select(-c(Kingdom:Genus))
  
  maas2 = maas %>% filter((Level=='Genus' & Genus %in% path.genus) | (Level=='Species' & Species %in% path.species)) %>% 
    mutate(Pathobiont = ifelse(is.na(Species),Genus,Species)) %>% 
    select(Pathobiont, qval) %>% 
    mutate(symb = ifelse(qval>0.05,'',ifelse(qval>0.01,"  * ",
                        ifelse(qval>0.001,' ** ',ifelse(qval>0.0001,' ***','****')))))
  
  # sample-specific data with qvals
  path=rbind(g,sp) %>% left_join(maas2)
  return(list(path=path,p=p,f=f))
}

path.data = load_path()
# Remove ns pathobiont
path = path.data$path %>% filter(symb != '')

# Contains all significant pathobionts, along with Proteobacteria and Enterobacteriaceae
taxa = full_join(path %>% rename(taxon=Pathobiont), path.data$p %>% rename(taxon=Phylum)) %>% 
  full_join(path.data$f %>% rename(taxon=Family)) %>% 
  filter(Day=="Day 28")

taxa$taxon = factor(taxa$taxon, levels = c('Proteobacteria','Enterobacteriaceae',unique(path$Pathobiont)))
```

# Correlate pathobionts with oxidative stress genes

```{r Combine}
temp = temp %>% select(Sample, Abundance, Diet, Day, Group, Subsystem.Level.3) %>% 
  rename(ab_ox = Abundance)
taxa = taxa %>% select(Sample, Abundance, Diet, Day, Group, taxon) %>% 
  rename(ab_tax = Abundance)

df = left_join(taxa,temp)

# Categorize pathobionts
df$cat = sapply(df$taxon,function(x)ifelse(x %in% c('Staphylococcus','Streptococcus','Prevotella copri'),'Down','Up'))
```

```{r Run stats and plot}
# CON
df.stats.con = df %>% filter(Diet=='CON',Day=='Day 28') %>% 
  group_by(Subsystem.Level.3,taxon,cat) %>% 
  group_modify(~cor.test(.$ab_ox,.$ab_tax,method='spearman')[c("statistic","p.value","estimate")] %>% as.data.frame()) %>% 
  ungroup %>% group_by(Subsystem.Level.3) %>% 
  mutate(qval = p.adjust(p.value,method='BH')) %>% ungroup %>% 
  mutate(Group = 'CON')
# LM
df.stats.lm = df %>% filter(Diet=='LM',Day=='Day 28') %>% 
  group_by(Subsystem.Level.3,taxon,cat) %>% 
  group_modify(~cor.test(.$ab_ox,.$ab_tax,method='spearman')[c("statistic","p.value","estimate")] %>% as.data.frame()) %>% 
  ungroup %>% group_by(Subsystem.Level.3) %>% 
  mutate(qval = p.adjust(p.value,method='BH')) %>% ungroup %>% 
  mutate(Group = 'LM')

df.stats = rbind(df.stats.con,df.stats.lm) %>% filter(!is.na(estimate)) %>% 
  mutate(label.lm = paste(Group,': \u03C1=',as.character(round(estimate,2)),', q=',signif(qval,2),sep='')) %>%
  mutate(label.con = ifelse(Group=='CON',label.lm,NA),
         label.lm = ifelse(Group=='LM',label.lm,NA))

# Arrange significance ~~~~~~~~~~~~~
temp = df.stats %>% filter(Group == 'CON',Subsystem.Level.3=='Oxidative stress', !(taxon %in% c('Proteobacteria','Enterobacteriaceae'))) %>% 
  arrange(estimate)
df.stats$taxon = factor(df.stats$taxon, levels = c('Proteobacteria','Enterobacteriaceae',as.character(temp$taxon)))

df.stats$Subsystem.Level.3[df.stats$Subsystem.Level.3=='Oxygen stress response / Human gut microbiome'] = 'Oxygen stress response/\nHuman gut microbiome'
df.stats$Subsystem.Level.3[df.stats$Subsystem.Level.3=='Regulation of Oxidative Stress Response'] = 'Regulation of\nOxidative Stress Response'

df.stats %>% 
  rename("Rho" = estimate) %>% 
  mutate(Qval = ifelse(qval>0.05,'',
                      ifelse(qval>0.01,'*',
                             ifelse(qval>0.001,'**',
                                    ifelse(qval<=0.001,'***',''))))) %>% 
  ggplot(aes(Subsystem.Level.3,taxon,col = Rho)) +
  geom_point(size =14) +
  theme_classic(base_size = 14) +
  scale_color_gradient2(low="darkblue", high="darkred", guide="colorbar") +
  geom_text(aes(label=Qval),size = 9, col = 'white',nudge_y = -0.15) +
  ylab(NULL) + xlab(NULL) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_grid(cols = vars(Group),rows = vars(cat), scales = 'free',space = 'free')
ggsave('Plots/Functional/SEED/P6_taxon_Heatmap_ox_stress.jpeg',height=8,width=6)
```

```{r Save stats}
write.csv(df.stats,'Results Files/Functional/Taxa_vs_Ox_Stress_Spearman.csv',row.names = F)
```

