---
title: "Fungal Analysis"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(phyloseq)
library(tidyverse)
library(ggpubr)
library(vegan)
library(Maaslin2)
```

# FUNGAL ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r}
# ps output is bacterial, raw counts
source('0. Initialize Fungal Data.R')

ps.rel = ps %>% microbiome::transform('compositional')

# Rarefied
set.seed(421)
ps.rar = ps %>% transform_sample_counts(function(x)round(x,0)) %>% rarefy_even_depth()
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Fig. 2a - Grouped Family Barplot

```{r}
# New phyloseq object:
family <- tax_glom(ps.rel, taxrank="Family", NArm=FALSE)
# Must be at least 1%
family2 = prune_taxa(taxa_sums(family)/nsamples(family) > 0.01, family) %>% 
  subset_taxa(!is.na(Family)) %>% 
  merge_samples('Group') # Proportions for each group now add up to 10 because n=10 for each group

temp = 10-phyloseq::sample_sums(family2) # Everything under 1% & unassigned reads
temp2 = otu_table(family2) %>% as.data.frame() %>% 
  cbind('Unknown/Other' = temp) %>% as.matrix()
temp2 = temp2/10 # Convert to average proportions /1
temp3 = tax_table(family2) %>% as.data.frame() %>% # Adjust the taxonomy table
  rbind('Unknown/Other' = c('Bacteria',rep('Unknown/Other',6))) %>% as.matrix()
sample_temp = sample_data(family2) %>% as.data.frame()
# Re-add this info, as it's corrupted by merge_samples
sample_temp$Diet[str_detect(row.names(sample_temp), 'CON')] = 'CON'
sample_temp$Diet[str_detect(row.names(sample_temp), 'LM')] = 'LM'
sample_temp$Day[str_detect(row.names(sample_temp), '0')] = 'Day 0'
sample_temp$Day[str_detect(row.names(sample_temp), '28')] = 'Day 28'
family3 <- phyloseq(otu_table(temp2, taxa_are_rows = F), sample_data(sample_temp), tax_table(temp3))

# Make plot
bb <- psmelt(family3)
bb_order <- c(unique(bb %>% filter(Family != 'Unknown/Other') %>% pull(Family)) %>% sort(),'Unknown/Other')
bb$Family<- factor(bb$Family, bb_order) #fill to total=1

p = ggplot(bb, aes(Day, Abundance, fill = Family))+
  geom_bar(stat = "identity", 
           position = "stack", 
           color = "black") +
  theme_classic(base_size = 18) +
  guides(fill = guide_legend(override.aes = list(colour = NULL), reverse=F)) + 
  theme(legend.key = element_rect(colour = "black")) + 
  theme(legend.position = "right",
        legend.text = element_text(size = 16)) + ylab('Rel. Abundance') + 
  facet_wrap('Diet') + xlab('')
data_2a = p$data %>% select(Sample,Diet,Day,Family,Abundance)

ggsave('Plots/Fungome/Fig. 2a - Grouped Family Barplot.png',height = 6, width = 9)
```

# Fig. 2a - Boxplots of Family-Level Sig Fungi

```{r Load Maaslin2 Stats}
# Maaslin stats
maas = readxl::read_xlsx('Results Files/Fungome/Fig. 2 Supplement - Fungal Maaslin2.xlsx',sheet='Family') %>% filter(qval<0.01)

ps.rel = ps %>% tax_glom('Family') %>% microbiome::transform('compositional') %>% 
  subset_taxa(Family %in% maas$Family) %>% psmelt() %>% 
  filter(Day=='Day 28') %>% 
  left_join(maas %>% select(Family,coef,qval)) %>% 
  mutate(symb = ifelse(qval>0.05,'',ifelse(qval>0.01,"  * ",
                      ifelse(qval>0.001,' ** ',ifelse(qval>0.0001,' ***','****')))))
```

```{r Create Boxplots}

# POSITIVE ASSOCIATIONS
# Set order
order = ps.rel %>% filter(Diet=='LM') %>% 
  group_by(Family) %>% 
  summarize(Abundance = mean(Abundance)) %>% ungroup() %>% 
  arrange(-Abundance)

# Set x and y coords for the p values
cols = c('CON' = '#FC746E','LM' = '#01BEC3')
max_per_path = ps.rel %>% group_by(Family) %>% summarize(M = 2*max(Abundance)) %>% ungroup %>% 
  mutate(Family = factor(Family, levels = order$Family))
max_pos = max_per_path %>%
  filter(Family %in% ps.rel$Family[ps.rel$coef>0]) %>% droplevels() %>% 
  arrange(Family) %>% mutate(N = c(1:nrow(.))+0.25)

# Positive assoc
temp = ps.rel %>% mutate(Family = factor(Family, levels = order$Family)) %>% 
  left_join(max_pos) %>% 
  filter(Family %in% ps.rel$Family[ps.rel$coef>0]) %>% droplevels() %>% 
  mutate(Abundance = Abundance + min(.$Abundance[.$Abundance !=0])/2) %>% 
  ggplot(aes(Family, Abundance, fill = Diet)) +
   geom_boxplot(outlier.shape = NA) +
   scale_fill_manual(values = cols) +
   geom_point(position = position_jitterdodge(jitter.height = 0, jitter.width = 0.2),size=1) +
   theme_classic(base_size = 18) + 
   guides(x = guide_axis(angle = 45)) +
   ylab('Rel Abundance') + xlab('') +
  geom_text(aes(x=N,y=M,label=symb), size = 6,lineheight=0.8,hjust=1) +
  scale_y_log10()

#~~~~~~~~~~~~~~~~~~~~~~~

# NEGATIVE ASSOCIATIONS
# Set order
order = ps.rel %>% filter(Diet=='CON') %>% 
  group_by(Family) %>% 
  summarize(Abundance = mean(Abundance)) %>% ungroup() %>% 
  arrange(-Abundance)

# Set x and y coords for the p values
cols = c('CON' = '#FC746E','LM' = '#01BEC3')
max_per_path = ps.rel %>% group_by(Family) %>% summarize(M = 2*max(Abundance)) %>% ungroup %>% 
  mutate(Family = factor(Family, levels = order$Family))
max_neg = max_per_path %>%
  filter(Family %in% ps.rel$Family[ps.rel$coef<0]) %>% droplevels() %>% 
  arrange(Family) %>% mutate(N = c(1:nrow(.))+0.25)

# Negative assoc
temp2 = ps.rel %>% mutate(Family = factor(Family, levels = order$Family)) %>% 
  left_join(max_neg) %>% 
  filter(Family %in% ps.rel$Family[ps.rel$coef<0]) %>% droplevels() %>% 
  mutate(Abundance = Abundance + min(.$Abundance[.$Abundance !=0])/2) %>% 
  ggplot(aes(Family, Abundance, fill = Diet)) +
   geom_boxplot(outlier.shape = NA) +
   scale_fill_manual(values = cols) +
   geom_point(position = position_jitterdodge(jitter.height = 0, jitter.width = 0.2),size=1) +
   theme_classic(base_size = 18) + 
   guides(x = guide_axis(angle = 45)) +
   ylab('Rel Abundance') + xlab('') +
  geom_text(aes(x=N,y=M,label=symb), size = 6,lineheight=0.8,hjust=1) +
  scale_y_log10()

ggarrange(plotlist = list(temp,temp2),ncol=2, common.legend = T, legend = 'right')
data_2a_increased = temp$data %>% select(Sample,Diet,Day,Family,Abundance)
data_2a_decreased = temp2$data %>% select(Sample,Diet,Day,Family,Abundance)

ggsave('Plots/Fungome/Fig. 2a - Family Boxplots.png', height = 5, width =10)
```

# Fig. 2b - Boxplots of Genus-Level Sig Fungi

```{r Load Maaslin2 Stats}
# Maaslin stats
maas = readxl::read_xlsx('Results Files/Fungome/Fig. 2 Supplement - Fungal Maaslin2.xlsx',sheet='Genus') %>% filter(qval<0.01)

ps.rel = ps %>% tax_glom('Genus') %>% microbiome::transform('compositional') %>% 
  subset_taxa(Genus %in% maas$Genus) %>% psmelt() %>% 
  filter(Day=='Day 28') %>% 
  left_join(maas %>% select(Genus,coef,qval)) %>% 
  mutate(symb = ifelse(qval>0.05,'',ifelse(qval>0.01,"  * ",
                      ifelse(qval>0.001,' ** ',ifelse(qval>0.0001,' ***','****')))))
```

```{r Create Boxplots}

# POSITIVE ASSOCIATIONS
# Set order
order = ps.rel %>% filter(Diet=='LM') %>% 
  group_by(Genus) %>% 
  summarize(Abundance = mean(Abundance)) %>% ungroup() %>% 
  arrange(-Abundance)

# Set x and y coords for the p values
cols = c('CON' = '#FC746E','LM' = '#01BEC3')
max_per_path = ps.rel %>% group_by(Genus) %>% summarize(M = 2*max(Abundance)) %>% ungroup %>% 
  mutate(Genus = factor(Genus, levels = order$Genus))
max_pos = max_per_path %>%
  filter(Genus %in% ps.rel$Genus[ps.rel$coef>0]) %>% droplevels() %>% 
  arrange(Genus) %>% mutate(N = c(1:nrow(.))+0.25)

# Positive assoc
temp = ps.rel %>% mutate(Genus = factor(Genus, levels = order$Genus)) %>% 
  left_join(max_pos) %>% 
  filter(Genus %in% ps.rel$Genus[ps.rel$coef>0]) %>% droplevels() %>% 
  mutate(Abundance = Abundance + min(.$Abundance[.$Abundance !=0])/2) %>% 
  ggplot(aes(Genus, Abundance, fill = Diet)) +
   geom_boxplot(outlier.shape = NA) +
   scale_fill_manual(values = cols) +
   geom_point(position = position_jitterdodge(jitter.height = 0, jitter.width = 0.2),size=1) +
   theme_classic(base_size = 18) + 
   guides(x = guide_axis(angle = 45)) +
   ylab('Rel Abundance') + xlab('') +
  geom_text(aes(x=N,y=M,label=symb), size = 6,lineheight=0.8,hjust=1) +
  scale_y_log10()

#~~~~~~~~~~~~~~~~~~~~~~~

# NEGATIVE ASSOCIATIONS
# Set order
order = ps.rel %>% filter(Diet=='CON') %>% 
  group_by(Genus) %>% 
  summarize(Abundance = mean(Abundance)) %>% ungroup() %>% 
  arrange(-Abundance)

# Set x and y coords for the p values
cols = c('CON' = '#FC746E','LM' = '#01BEC3')
max_per_path = ps.rel %>% group_by(Genus) %>% summarize(M = 2*max(Abundance)) %>% ungroup %>% 
  mutate(Genus = factor(Genus, levels = order$Genus))
max_neg = max_per_path %>%
  filter(Genus %in% ps.rel$Genus[ps.rel$coef<0]) %>% droplevels() %>% 
  arrange(Genus) %>% mutate(N = c(1:nrow(.))+0.25)

# Negative assoc
temp2 = ps.rel %>% mutate(Genus = factor(Genus, levels = order$Genus)) %>% 
  left_join(max_neg) %>% 
  filter(Genus %in% ps.rel$Genus[ps.rel$coef<0]) %>% droplevels() %>% 
  mutate(Abundance = Abundance + min(.$Abundance[.$Abundance !=0])/2) %>% 
  ggplot(aes(Genus, Abundance, fill = Diet)) +
   geom_boxplot(outlier.shape = NA) +
   scale_fill_manual(values = cols) +
   geom_point(position = position_jitterdodge(jitter.height = 0, jitter.width = 0.2),size=1) +
   theme_classic(base_size = 18) + 
   guides(x = guide_axis(angle = 45)) +
   ylab('Rel Abundance') + xlab('') +
  geom_text(aes(x=N,y=M,label=symb), size = 6,lineheight=0.8,hjust=1) +
  scale_y_log10()

ggarrange(plotlist = list(temp,temp2),ncol=2, common.legend = T, legend = 'right')
data_2b_increased = temp$data %>% select(Sample,Diet,Day,Family,Abundance)
data_2b_decreased = temp2$data %>% select(Sample,Diet,Day,Family,Abundance)

ggsave('Plots/Fungome/Fig. 2b - Genus Boxplots.png', height = 5, width =10)
```

# Fig. 2c - Alpha Diversity

```{r}
otu = ps.rar@otu_table %>% as.matrix %>% as.data.frame
# Use Vegan package to calculate diversity
df_alpha <- diversity(t(otu), index='shannon', MARGIN = 1, base = exp(1)) %>%
  as.data.frame()
df_alpha$Sample = rownames(df_alpha)
df_alpha = df_alpha %>% left_join(sample_data(ps.rar) %>% as.matrix() %>% 
                                    as.data.frame %>% select(Diet, Day) %>% rownames_to_column('Sample'))
names(df_alpha)[1] = 'shannon'
  
# Plot results - samples are paired
p=ggplot(df_alpha, aes(x=Diet, y=shannon)) + 
      geom_boxplot(outlier.shape = NA) +
      geom_jitter(aes(color=Day),  height = 0, width = 0.2, size=3) + 
      theme_classic(base_size = 16) +
      stat_compare_means(vjust=0.5, label.y = 5, paired=T, size = 5)+
      ylab('Shannon Diversity') +  xlab('') +
      facet_wrap(~Day) +
      theme(legend.position = 'none')
data_2c = p$data %>% select(Sample,Diet,Day,shannon)
ggsave('Plots/Fungome/Fig. 2c - Alpha Diversity.png',height = 3.5, width = 6)
```

# Fig. 2d - Beta Diversity

```{r Calculate Beta diversity}
# Calculate the distance matrix:
beta_dist <- vegdist(t(otu), index = 'bray')
# Scaling results with NMDS
mds <- metaMDS(beta_dist)
# Extract PCoA axes and format
mds_data <- as.data.frame(mds$points)
mds_data$`Sample` <- rownames(mds_data)
mds_data <- mds_data %>% left_join(sample_data(ps.rar) %>% as.matrix() %>% 
                                    as.data.frame %>% select(Diet, Day) %>% rownames_to_column('Sample')) %>% 
  mutate(beta_type = 'Bray-Curtis')

# Plot results
mds_data$Group = paste(mds_data$Diet, mds_data$Day)
mds_data$Group = factor(mds_data$Group, 
                        levels = c('CON Day 0','LM Day 0','CON Day 28','LM Day 28'))
g=ggplot(mds_data, aes(x = MDS1, y = MDS2, color = Group)) +
    geom_point(size = 2) + 
    stat_ellipse(size = 1) +
    theme_classic(base_size = 20) + 
    xlab('PCoA 1') + ylab('PCoA 2') +
    ggtitle('Bray-Curtis Beta Diversity')
data_2d = g$data
ggsave('Plots/Fungome/Fig. 2d - Beta Diversity.png',height = 5, width = 7)
```

```{r Calculate Beta statistics}
# CENTROID DISTANCES (PERMANOVA)~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Format datasets
count_mat = otu %>% t() %>% as.data.frame() %>% rownames_to_column('Sample') %>% 
  left_join(mds_data %>% select(Sample, Diet, Day, Group))
      
# CON D0 vs D28
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'CON')]
cm_con = otu %>% t %>% .[str_detect(rownames(.),'CON'),]
beta_dist_con <- vegdist(cm_con, index = 'bray')
pathotype.adonis.con <- ((adonis(beta_dist_con ~ temp_group))$aov.tab)[1,6] # p value
      
# LM D0 vs D28
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'LM')]
cm_lm = otu %>% t %>% .[str_detect(rownames(.),'LM'),]
beta_dist_lm <- vegdist(cm_lm, index = 'bray')
pathotype.adonis.lm <- ((adonis(beta_dist_lm ~ temp_group))$aov.tab)[1,6] # p value
      
# D0 CON vs LM
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'D0')]
cm_d0 = otu %>% t %>% .[str_detect(rownames(.),'D0'),]
beta_dist_d0 <- vegdist(cm_d0, index = 'bray')
pathotype.adonis.d0 <- ((adonis(beta_dist_d0 ~ temp_group))$aov.tab)[1,6] # p value

# D28 CON vs LM
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'D28')]
cm_d28 = otu %>% t %>% .[str_detect(rownames(.),'D28'),]
beta_dist_d28 <- vegdist(cm_d28, index = 'bray')
pathotype.adonis.d28 <- ((adonis(beta_dist_d28 ~ temp_group))$aov.tab)[1,6] # p value

# CONCATENATE STATISTICS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

beta_stats = tibble(Stat = rep('Centroid Distances',4),
                    Comparison = c('CON D0 vs D28','LM D0 vs D28',
                                   'D0 CON vs LM','D28 CON vs LM'),
                    pval = c(pathotype.adonis.con, pathotype.adonis.lm,
                             pathotype.adonis.d0, pathotype.adonis.d28)) %>% 
  mutate(`p adj` = p.adjust(pval)) %>% 
  mutate(`Beta Measure` = 'Bray-Curtis') %>% 
  select(`Beta Measure`, Stat, everything()) %>% 
  mutate(pval = round(pval,4),`p adj` = round(`p adj`,4))

write.csv(beta_stats,'Results Files/Fungome/Beta Diversity Statistics.csv',row.names = F)

rm(beta_dist,beta_dist_con,beta_dist_d0,beta_dist_d28,beta_dist_lm,
   cm_con,cm_lm,cm_d0,cm_d28,
   pathotype.adonis.con,pathotype.adonis.d0,pathotype.adonis.d28,
   pathotype.adonis.lm,pathotype.disp,pathotype.disp.TukeyHSD,
   count_mat, count_mat_2, mds, mds_data, temp_group)
```
# VIRAL ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```{r}
# ps output is bacterial, raw counts
source('0. Initialize Viral Data.R')

ps.rel = ps %>% microbiome::transform('compositional')

# Rarefied
set.seed(421)
ps.rar = ps %>% transform_sample_counts(function(x)round(x,0)) %>% rarefy_even_depth()
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Fig. 2e - Grouped Family Barplot

```{r}
# New phyloseq object:
family <- tax_glom(ps.rel, taxrank="Family", NArm=FALSE)
# Must be at least 1%
family2 = prune_taxa(taxa_sums(family)/nsamples(family) > 0.01, family) %>% 
  subset_taxa(!is.na(Family)) %>% 
  merge_samples('Group') # Proportions for each group now add up to 10 because n=10 for each group

temp = 10-phyloseq::sample_sums(family2) # Everything under 1% & unassigned reads
temp2 = otu_table(family2) %>% as.data.frame() %>% 
  cbind('Unknown/Other' = temp) %>% as.matrix()
temp2 = temp2/10 # Convert to average proportions /1
temp3 = tax_table(family2) %>% as.data.frame() %>% # Adjust the taxonomy table
  rbind('Unknown/Other' = c('Bacteria',rep('Unknown/Other',6))) %>% as.matrix()
sample_temp = sample_data(family2) %>% as.data.frame()
# Re-add this info, as it's corrupted by merge_samples
sample_temp$Diet[str_detect(row.names(sample_temp), 'CON')] = 'CON'
sample_temp$Diet[str_detect(row.names(sample_temp), 'LM')] = 'LM'
sample_temp$Day[str_detect(row.names(sample_temp), '0')] = 'Day 0'
sample_temp$Day[str_detect(row.names(sample_temp), '28')] = 'Day 28'
family3 <- phyloseq(otu_table(temp2, taxa_are_rows = F), sample_data(sample_temp), tax_table(temp3))

# Make plot
bb <- psmelt(family3)
bb_order <- c(unique(bb %>% filter(Family != 'Unknown/Other') %>% pull(Family)) %>% sort(),'Unknown/Other')
bb$Family<- factor(bb$Family, bb_order) #fill to total=1

p = ggplot(bb, aes(Day, Abundance, fill = Family))+
  geom_bar(stat = "identity", 
           position = "stack", 
           color = "black") +
  theme_classic(base_size = 18) +
  guides(fill = guide_legend(override.aes = list(colour = NULL), reverse=F)) + 
  theme(legend.key = element_rect(colour = "black")) + 
  theme(legend.position = "right",
        legend.text = element_text(size = 16)) + ylab('Rel. Abundance') + 
  facet_wrap('Diet') + xlab('')
data_2e = p$data %>% select(Sample,Diet,Day,Family,Abundance)

ggsave('Plots/Virome/Fig. 2e - Grouped Family Barplot.png',height = 6, width = 9)
```

# Fig. 2e - Boxplots of Family-Level Sig Fungi

```{r Load Maaslin2 Stats}
# Maaslin stats
maas = readxl::read_xlsx('Results Files/Virome/Fig. 2 Supplement - Viral Maaslin2.xlsx',sheet='Family') %>% filter(qval<0.01)

ps.rel = ps %>% tax_glom('Family') %>% microbiome::transform('compositional') %>% 
  subset_taxa(Family %in% maas$Family) %>% psmelt() %>% 
  filter(Day=='Day 28') %>% 
  left_join(maas %>% select(Family,coef,qval)) %>% 
  mutate(symb = ifelse(qval>0.05,'',ifelse(qval>0.01,"  * ",
                      ifelse(qval>0.001,' ** ',ifelse(qval>0.0001,' ***','****')))))
```

```{r Create Boxplots}

# POSITIVE ASSOCIATIONS
# Set order
order = ps.rel %>% filter(Diet=='LM') %>% 
  group_by(Family) %>% 
  summarize(Abundance = mean(Abundance)) %>% ungroup() %>% 
  arrange(-Abundance)

# Set x and y coords for the p values
cols = c('CON' = '#FC746E','LM' = '#01BEC3')
max_per_path = ps.rel %>% group_by(Family) %>% summarize(M = 2*max(Abundance)) %>% ungroup %>% 
  mutate(Family = factor(Family, levels = order$Family))
max_pos = max_per_path %>%
  filter(Family %in% ps.rel$Family[ps.rel$coef>0]) %>% droplevels() %>% 
  arrange(Family) %>% mutate(N = c(1:nrow(.))+0.25)

# Positive assoc
temp = ps.rel %>% mutate(Family = factor(Family, levels = order$Family)) %>% 
  left_join(max_pos) %>% 
  filter(Family %in% ps.rel$Family[ps.rel$coef>0]) %>% droplevels() %>% 
  mutate(Abundance = Abundance + min(.$Abundance[.$Abundance !=0])/2) %>% 
  ggplot(aes(Family, Abundance, fill = Diet)) +
   geom_boxplot(outlier.shape = NA) +
   scale_fill_manual(values = cols) +
   geom_point(position = position_jitterdodge(jitter.height = 0, jitter.width = 0.2),size=1) +
   theme_classic(base_size = 18) + 
   guides(x = guide_axis(angle = 45)) +
   ylab('Rel Abundance') + xlab('') +
  geom_text(aes(x=N,y=M,label=symb), size = 6,lineheight=0.8,hjust=1) +
  scale_y_log10()

#~~~~~~~~~~~~~~~~~~~~~~~

# NEGATIVE ASSOCIATIONS
# Set order
order = ps.rel %>% filter(Diet=='CON') %>% 
  group_by(Family) %>% 
  summarize(Abundance = mean(Abundance)) %>% ungroup() %>% 
  arrange(-Abundance)

# Set x and y coords for the p values
cols = c('CON' = '#FC746E','LM' = '#01BEC3')
max_per_path = ps.rel %>% group_by(Family) %>% summarize(M = 2*max(Abundance)) %>% ungroup %>% 
  mutate(Family = factor(Family, levels = order$Family))
max_neg = max_per_path %>%
  filter(Family %in% ps.rel$Family[ps.rel$coef<0]) %>% droplevels() %>% 
  arrange(Family) %>% mutate(N = c(1:nrow(.))+0.25)

# Negative assoc
temp2 = ps.rel %>% mutate(Family = factor(Family, levels = order$Family)) %>% 
  left_join(max_neg) %>% 
  filter(Family %in% ps.rel$Family[ps.rel$coef<0]) %>% droplevels() %>% 
  mutate(Abundance = Abundance + min(.$Abundance[.$Abundance !=0])/2) %>% 
  ggplot(aes(Family, Abundance, fill = Diet)) +
   geom_boxplot(outlier.shape = NA) +
   scale_fill_manual(values = cols) +
   geom_point(position = position_jitterdodge(jitter.height = 0, jitter.width = 0.2),size=1) +
   theme_classic(base_size = 18) + 
   guides(x = guide_axis(angle = 45)) +
   ylab('Rel Abundance') + xlab('') +
  geom_text(aes(x=N,y=M,label=symb), size = 6,lineheight=0.8,hjust=1) +
  scale_y_log10()

ggarrange(plotlist = list(temp,temp2),ncol=2, widths = c(3,2.3), common.legend = T, legend = 'right')
data_2e_increased = temp$data %>% select(Sample,Diet,Day,Family,Abundance)
data_2e_decreased = temp2$data %>% select(Sample,Diet,Day,Family,Abundance)

ggsave('Plots/Virome/Fig. 2e - Family Boxplots.png', height = 5, width =10)
```

# Fig. 2f - Beta Diversity

```{r Calculate Beta diversity}
otu = ps.rar@otu_table %>% as.matrix %>% as.data.frame
# Calculate the distance matrix:
beta_dist <- vegdist(t(otu), index = 'bray')
# Scaling results with NMDS
mds <- metaMDS(beta_dist)
# Extract PCoA axes and format
mds_data <- as.data.frame(mds$points)
mds_data$`Sample` <- rownames(mds_data)
mds_data <- mds_data %>% left_join(sample_data(ps.rar) %>% as.matrix() %>% 
                                    as.data.frame %>% select(Diet, Day) %>% rownames_to_column('Sample')) %>% 
  mutate(beta_type = 'Bray-Curtis')

# Plot results
mds_data$Group = paste(mds_data$Diet, mds_data$Day)
mds_data$Group = factor(mds_data$Group, 
                        levels = c('CON Day 0','LM Day 0','CON Day 28','LM Day 28'))
g=ggplot(mds_data, aes(x = MDS1, y = MDS2, color = Group)) +
    geom_point(size = 2) + 
    stat_ellipse(size = 1) +
    theme_classic(base_size = 20) + 
    xlab('PCoA 1') + ylab('PCoA 2') +
    ggtitle('Bray-Curtis Beta Diversity')
data_2f = g$data

ggsave('Plots/Virome/Fig. 2f - Beta Diversity.png',height = 5, width = 7)
```

```{r Calculate Beta statistics}
# CENTROID DISTANCES (PERMANOVA)~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Format datasets
count_mat = otu %>% t() %>% as.data.frame() %>% rownames_to_column('Sample') %>% 
  left_join(mds_data %>% select(Sample, Diet, Day, Group))
      
# CON D0 vs D28
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'CON')]
cm_con = otu %>% t %>% .[str_detect(rownames(.),'CON'),]
beta_dist_con <- vegdist(cm_con, index = 'bray')
pathotype.adonis.con <- ((adonis(beta_dist_con ~ temp_group))$aov.tab)[1,6] # p value
      
# LM D0 vs D28
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'LM')]
cm_lm = otu %>% t %>% .[str_detect(rownames(.),'LM'),]
beta_dist_lm <- vegdist(cm_lm, index = 'bray')
pathotype.adonis.lm <- ((adonis(beta_dist_lm ~ temp_group))$aov.tab)[1,6] # p value
      
# D0 CON vs LM
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'D0')]
cm_d0 = otu %>% t %>% .[str_detect(rownames(.),'D0'),]
beta_dist_d0 <- vegdist(cm_d0, index = 'bray')
pathotype.adonis.d0 <- ((adonis(beta_dist_d0 ~ temp_group))$aov.tab)[1,6] # p value

# D28 CON vs LM
temp_group = count_mat$Group[str_detect(count_mat$Sample, 'D28')]
cm_d28 = otu %>% t %>% .[str_detect(rownames(.),'D28'),]
beta_dist_d28 <- vegdist(cm_d28, index = 'bray')
pathotype.adonis.d28 <- ((adonis(beta_dist_d28 ~ temp_group))$aov.tab)[1,6] # p value

# CONCATENATE STATISTICS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

beta_stats = tibble(Stat = rep('Centroid Distances',4),
                    Comparison = c('CON D0 vs D28','LM D0 vs D28',
                                   'D0 CON vs LM','D28 CON vs LM'),
                    pval = c(pathotype.adonis.con, pathotype.adonis.lm,
                             pathotype.adonis.d0, pathotype.adonis.d28)) %>% 
  mutate(`p adj` = p.adjust(pval)) %>% 
  mutate(`Beta Measure` = 'Bray-Curtis') %>% 
  select(`Beta Measure`, Stat, everything()) %>% 
  mutate(pval = round(pval,4),`p adj` = round(`p adj`,4))

write.csv(beta_stats,'Results Files/Virome/Beta Diversity Statistics.csv',row.names = F)

rm(beta_dist,beta_dist_con,beta_dist_d0,beta_dist_d28,beta_dist_lm,
   cm_con,cm_lm,cm_d0,cm_d28,
   pathotype.adonis.con,pathotype.adonis.d0,pathotype.adonis.d28,
   pathotype.adonis.lm,
   count_mat, mds, mds_data, temp_group)
```

# Save Source Data

```{r}
L = list('Fig 2a' = data_2a, 'Fig 2a Inc' = data_2a_increased, 'Fig 2a Dec' = data_2a_decreased,
         'Fig 2b Inc' = data_2b_increased, 'Fig 2b Dec' = data_2b_decreased,
         'Fig 2c' = data_2c, 'Fig 2d' = data_2d, 
         'Fig 2e Barplot' = data_2e, 'Fig 2e Inc' = data_2e_increased, 'Fig 2e Dec' = data_2e_decreased,
         'Fig 2f' = data_2f)
writexl::write_xlsx(L,'Source Data/Source Data Fig 2.xlsx')
```