---
title: "Fungal Analysis"
author: "Avril Metcalfe-Roach"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(phyloseq)
library(tidyverse)
library(ggpubr)
library(vegan)
library(Maaslin2)

# ps output is raw counts
ps = readRDS('Data Files/bacterial_phyloseq.rds')

ps.rel = ps %>% microbiome::transform('compositional')

# Rarefied
set.seed(421)
ps.rar = ps %>% transform_sample_counts(function(x)round(x,0)) %>% rarefy_even_depth()
```

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Grouped Barplots

```{r S1a - Family}
# New phyloseq object:
family <- tax_glom(ps.rel, taxrank="Family", NArm=FALSE)
# Must be at least 1%
family2 = prune_taxa(taxa_sums(family)/nsamples(family) > 0.01, family) %>% 
  subset_taxa(!is.na(Family)) %>% 
  merge_samples('Group') # Proportions for each group now add up to 10 because n=10 for each group

temp = 10-phyloseq::sample_sums(family2) # Everything under 1% & unassigned reads
temp2 = otu_table(family2) %>% as.data.frame() %>% 
  cbind('Unknown/Other' = temp) %>% as.matrix()
temp2 = temp2/10 # Convert to average proportions /1
temp3 = tax_table(family2) %>% as.data.frame() %>% # Adjust the taxonomy table
  rbind('Unknown/Other' = c('Bacteria',rep('Unknown/Other',6))) %>% as.matrix()
sample_temp = sample_data(family2) %>% as.data.frame()
# Re-add this info, as it's corrupted by merge_samples
sample_temp$Diet[str_detect(row.names(sample_temp), 'CON')] = 'CON'
sample_temp$Diet[str_detect(row.names(sample_temp), 'LM')] = 'LM'
sample_temp$Day[str_detect(row.names(sample_temp), '0')] = 'Day 0'
sample_temp$Day[str_detect(row.names(sample_temp), '28')] = 'Day 28'
family3 <- phyloseq(otu_table(temp2, taxa_are_rows = F), sample_data(sample_temp), tax_table(temp3))

# Make plot
bb <- psmelt(family3)
bb_order <- c(unique(bb %>% filter(Family != 'Unknown/Other') %>% pull(Family)) %>% sort(),'Unknown/Other')
bb$Family<- factor(bb$Family, bb_order) #fill to total=1

p = ggplot(bb, aes(Day, Abundance, fill = Family))+
  geom_bar(stat = "identity", 
           position = "stack", 
           color = "black") +
  theme_classic(base_size = 18) +
  guides(fill = guide_legend(override.aes = list(colour = NULL), reverse=F)) + 
  theme(legend.key = element_rect(colour = "black")) + 
  theme(legend.position = "right",
        legend.text = element_text(size = 16)) + ylab('Rel. Abundance') + 
  facet_wrap('Diet') + xlab('')

ggsave('Plots/Bacteriome/Fig. S1a - Grouped Family Barplot.png',height = 6, width = 9)
```

```{r S1b - Genus}
# New phyloseq object:
genus <- tax_glom(ps.rel, taxrank="Genus", NArm=FALSE)
# Must be at least 1%
genus2 = prune_taxa(taxa_sums(genus)/nsamples(genus) > 0.01, genus) %>% 
  subset_taxa(!is.na(Genus)) %>% 
  merge_samples('Group') # Proportions for each group now add up to 10 because n=10 for each group

temp = 10-phyloseq::sample_sums(genus2) # Everything under 1% & unassigned reads
temp2 = otu_table(genus2) %>% as.data.frame() %>% 
  cbind('Unknown/Other' = temp) %>% as.matrix()
temp2 = temp2/10 # Convert to average proportions /1
temp3 = tax_table(genus2) %>% as.data.frame() %>% # Adjust the taxonomy table
  rbind('Unknown/Other' = c('Bacteria',rep('Unknown/Other',6))) %>% as.matrix()
sample_temp = sample_data(genus2) %>% as.data.frame()
# Re-add this info, as it's corrupted by merge_samples
sample_temp$Diet[str_detect(row.names(sample_temp), 'CON')] = 'CON'
sample_temp$Diet[str_detect(row.names(sample_temp), 'LM')] = 'LM'
sample_temp$Day[str_detect(row.names(sample_temp), '0')] = 'Day 0'
sample_temp$Day[str_detect(row.names(sample_temp), '28')] = 'Day 28'
genus3 <- phyloseq(otu_table(temp2, taxa_are_rows = F), sample_data(sample_temp), tax_table(temp3))

# Make plot
bb <- psmelt(genus3)
bb_order <- c(unique(bb %>% filter(Genus != 'Unknown/Other') %>% pull(Genus)) %>% sort(),'Unknown/Other')
bb$Genus<- factor(bb$Genus, bb_order) #fill to total=1

p = ggplot(bb, aes(Day, Abundance, fill = Genus))+
  geom_bar(stat = "identity", 
           position = "stack", 
           color = "black") +
  theme_classic(base_size = 18) +
  guides(fill = guide_legend(override.aes = list(colour = NULL), reverse=F)) + 
  theme(legend.key = element_rect(colour = "black")) + 
  theme(legend.position = "right",
        legend.text = element_text(size = 16)) + ylab('Rel. Abundance') + 
  facet_wrap('Diet') + xlab('')

ggsave('Plots/Bacteriome/Fig. S1b - Grouped Genus Barplot.png',height = 6, width = 9)
```

```{r S1c - Species}
# New phyloseq object:
species <- ps.rel
# Must be at least 1%
species2 = prune_taxa(taxa_sums(species)/nsamples(species) > 0.01, species) %>% 
  subset_taxa(!is.na(Species)) %>% 
  merge_samples('Group') # Proportions for each group now add up to 10 because n=10 for each group

temp = 10-phyloseq::sample_sums(species2) # Everything under 1% & unassigned reads
temp2 = otu_table(species2) %>% as.data.frame() %>% 
  cbind('Unknown/Other' = temp) %>% as.matrix()
temp2 = temp2/10 # Convert to average proportions /1
temp3 = tax_table(species2) %>% as.data.frame() %>% # Adjust the taxonomy table
  rbind('Unknown/Other' = c('Bacteria',rep('Unknown/Other',6))) %>% as.matrix()
sample_temp = sample_data(species2) %>% as.data.frame()
# Re-add this info, as it's corrupted by merge_samples
sample_temp$Diet[str_detect(row.names(sample_temp), 'CON')] = 'CON'
sample_temp$Diet[str_detect(row.names(sample_temp), 'LM')] = 'LM'
sample_temp$Day[str_detect(row.names(sample_temp), '0')] = 'Day 0'
sample_temp$Day[str_detect(row.names(sample_temp), '28')] = 'Day 28'
species3 <- phyloseq(otu_table(temp2, taxa_are_rows = F), sample_data(sample_temp), tax_table(temp3))

# Make plot
bb <- psmelt(species3) %>% 
  mutate(Species = str_wrap(Species,40))
bb_order <- c(unique(bb %>% filter(Species != 'Unknown/Other') %>% pull(Species)) %>% sort(),'Unknown/Other')
bb$Species<- factor(bb$Species, bb_order) #fill to total=1

p = ggplot(bb, aes(Day, Abundance, fill = Species))+
  geom_bar(stat = "identity", 
           position = "stack", 
           color = "black") +
  theme_classic(base_size = 18) +
  guides(fill = guide_legend(override.aes = list(colour = NULL), reverse=F)) + 
  theme(legend.key = element_rect(colour = "black")) + 
  theme(legend.position = "right",
        legend.text = element_text(size = 16)) + ylab('Rel. Abundance') + 
  facet_wrap('Diet') + xlab('')

ggsave('Plots/Bacteriome/Fig. S1c - Grouped Species Barplot.png',height = 6, width = 12)
```

# Individual Phylum Barplots

```{r S1d - Family}
# New phyloseq object:
family <- tax_glom(ps.rel, taxrank="Family", NArm=FALSE)
# Must be at least 1%
family2 = prune_taxa(taxa_sums(family)/nsamples(family) > 0.01, family) %>% 
  subset_taxa(!is.na(Family))

# Adjust phyloseq to group everything under 1%
temp = 1-phyloseq::sample_sums(family2) # Everything under 1% & unassigned reads
temp2 = otu_table(family2) %>% as.data.frame() %>% 
  rbind('Unknown/Other' = temp) %>% as.matrix()
temp3 = tax_table(family2) %>% as.data.frame() %>% # Adjust the taxonomy table
  rbind('Unknown/Other' = c('Bacteria',rep('Unknown/Other',6))) %>% as.matrix()
# Extract mouse numbers
sample_temp = sample_data(family2) %>% as.data.frame()
sample_temp$Mouse = sapply(sample_temp$Sample,function(x) x %>% str_split('_') %>% .[[1]] %>% .[2])
family3 <- phyloseq(otu_table(temp2, taxa_are_rows = T), sample_data(sample_temp), tax_table(temp3))

# Make plot
bb <- psmelt(family3)
bb_order <- c(unique(bb %>% filter(Family != 'Unknown/Other') %>% pull(Family)) %>% sort(),'Unknown/Other')
bb$Family<- factor(bb$Family, bb_order) #fill to total=1

p = ggplot(bb, aes(Mouse, Abundance, fill = Family))+
  geom_bar(stat = "identity", 
           position = "stack", 
           color = "black") +
  theme_classic(base_size = 18) +
  guides(fill = guide_legend(override.aes = list(colour = NULL), reverse=F)) + 
  theme(legend.key = element_rect(colour = "black")) + 
  theme(legend.position = "right",
        legend.text = element_text(size = 16)) + ylab('Rel. Abundance') + 
  facet_wrap('Group')

ggsave('Plots/Bacteriome/Fig. S1d - Individual Family Barplots.png',height = 8, width = 10)
```

```{r S1d - Genus}
# New phyloseq object:
genus <- tax_glom(ps.rel, taxrank="Genus", NArm=FALSE)
# Must be at least 1%
genus2 = prune_taxa(taxa_sums(genus)/nsamples(genus) > 0.01, genus) %>% 
  subset_taxa(!is.na(Genus))

# Adjust phyloseq to group everything under 1%
temp = 1-phyloseq::sample_sums(genus2) # Everything under 1% & unassigned reads
temp2 = otu_table(genus2) %>% as.data.frame() %>% 
  rbind('Unknown/Other' = temp) %>% as.matrix()
temp3 = tax_table(genus2) %>% as.data.frame() %>% # Adjust the taxonomy table
  rbind('Unknown/Other' = c('Bacteria',rep('Unknown/Other',6))) %>% as.matrix()
# Extract mouse numbers
sample_temp = sample_data(genus2) %>% as.data.frame()
sample_temp$Mouse = sapply(sample_temp$Sample,function(x) x %>% str_split('_') %>% .[[1]] %>% .[2])
genus3 <- phyloseq(otu_table(temp2, taxa_are_rows = T), sample_data(sample_temp), tax_table(temp3))

# Make plot
bb <- psmelt(genus3)
bb_order <- c(unique(bb %>% filter(Genus != 'Unknown/Other') %>% pull(Genus)) %>% sort(),'Unknown/Other')
bb$Genus<- factor(bb$Genus, bb_order) #fill to total=1

p = ggplot(bb, aes(Mouse, Abundance, fill = Genus))+
  geom_bar(stat = "identity", 
           position = "stack", 
           color = "black") +
  theme_classic(base_size = 18) +
  guides(fill = guide_legend(override.aes = list(colour = NULL), reverse=F)) + 
  theme(legend.key = element_rect(colour = "black")) + 
  theme(legend.position = "right",
        legend.text = element_text(size = 16)) + ylab('Rel. Abundance') + 
  facet_wrap('Group')

ggsave('Plots/Bacteriome/Fig. S1e - Individual Genus Barplots.png',height = 8, width = 10)
```

```{r S1d - Species}
# New phyloseq object:
species <- ps.rel
# Must be at least 1%
species2 = prune_taxa(taxa_sums(species)/nsamples(species) > 0.01, species) %>% 
  subset_taxa(!is.na(Species))

# Adjust phyloseq to group everything under 1%
temp = 1-phyloseq::sample_sums(species2) # Everything under 1% & unassigned reads
temp2 = otu_table(species2) %>% as.data.frame() %>% 
  rbind('Unknown/Other' = temp) %>% as.matrix()
temp3 = tax_table(species2) %>% as.data.frame() %>% # Adjust the taxonomy table
  rbind('Unknown/Other' = c('Bacteria',rep('Unknown/Other',6))) %>% as.matrix()
# Extract mouse numbers
sample_temp = sample_data(species2) %>% as.data.frame()
sample_temp$Mouse = sapply(sample_temp$Sample,function(x) x %>% str_split('_') %>% .[[1]] %>% .[2])
species3 <- phyloseq(otu_table(temp2, taxa_are_rows = T), sample_data(sample_temp), tax_table(temp3))

# Make plot
bb <- psmelt(species3) %>% 
  mutate(Species = str_wrap(Species,40))
bb_order <- c(unique(bb %>% filter(Species != 'Unknown/Other') %>% pull(Species)) %>% sort(),'Unknown/Other')
bb$Species<- factor(bb$Species, bb_order) #fill to total=1

p = ggplot(bb, aes(Mouse, Abundance, fill = Species))+
  geom_bar(stat = "identity", 
           position = "stack", 
           color = "black") +
  theme_classic(base_size = 18) +
  guides(fill = guide_legend(override.aes = list(colour = NULL), reverse=F)) + 
  theme(legend.key = element_rect(colour = "black")) + 
  theme(legend.position = "right",
        legend.text = element_text(size = 16)) + ylab('Rel. Abundance') + 
  facet_wrap('Group')

ggsave('Plots/Bacteriome/Fig. S1f - Individual Species Barplots.png',height = 8, width = 12)
```
